### PR TITLE
osd/scrub: separate between PG state flags and internal scrubber operation

### PR DESCRIPTION
**Pull request by ronen-fr**
Modify the scrubber to rely on internal flags for 'should we repair' and
'is this a deep scrub', instead of using PG_STATE_REPAIR & PG_STATE_DEEP_SCRUB.

This enables us to implement the 'fix-as-you-go deepscrub' functionality
of 'osd_scrub_auto_repair', without displaying REPAIR status to the user.

Fixes: https://tracker.ceph.com/issues/50446

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>
Co-Authored by: Neha Ojha <nojha@redhat.com>

### CODE REVIEW

File: src/osd/pg_scrubber.cc [hunk 1]
```diff
@@ -624,7 +625,7 @@ void PgScrubber::add_delayed_scheduling()
 
   milliseconds sleep_time{0ms};
   if (m_needs_sleep) {
-    double scrub_sleep = 1000.0 * m_osds->osd->scrub_sleep_time(m_flags.required);
+    double scrub_sleep = 700.0 + 1000.0 * m_osds->osd->scrub_sleep_time(m_flags.required);
```
Comment (ronen-fr) [inline src/osd/pg_scrubber.cc:628]: There seems to be a problem with the value of scrub_sleep_time conf value.
Looking into it now.
The +700 hack is needed to pass minimal standalone testing

File: src/osd/pg_scrubber.cc [hunk 2]
```diff
@@ -1133,8 +1144,15 @@ void PgScrubber::set_op_parameters(requested_scrub_t& request)
     state_set(PG_STATE_DEEP_SCRUB);
   }
 
-  if (request.must_repair || m_flags.auto_repair) {
+  // m_is_repair is set for either 'must_repair' or 'repair-on-the-go' (i.e.
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/pg_scrubber.cc:1147]: The problem is that auto-repair is already an existing term, which is also set for the deep-scrub that was initiated
after a problem was detected. All in all - there are too many 'auto-something' flags and internal statuses (too many for
me to remember without consulting the code).

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-12_04:48:43-rados-wip-kefu-testing2-2021-05-12-1103-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-13_03:34:40-rados-wip-kefu-testing-2021-05-12-2313-distro-basic-smithi/

**Comment by ronen-fr** [comment]
A test added to qa/standalone/scrub/osd-scrub-repair.sh (as a separate commit).

File: qa/standalone/scrub/osd-scrub-repair.sh [hunk 3]
```diff
@@ -329,6 +329,126 @@ function TEST_auto_repair_erasure_coded_overwrites() {
...
+function wait_end_of_scrub() { # osd# pg
+    local the_osd="osd.$1"
+    local pgid=$2
+    for ((i=0; i < 20; i++)); do
+
+        st=`ceph pg $pgid query --format json | jq '.state' `
+        echo "RRR wait-scrub-end state now: " $st
+        [[ $st =~ (.*scrubbing.*) ]] || break
+        flush_pg_stats
+    else
```
Comment (neha-ojha) [inline qa/standalone/scrub/osd-scrub-repair.sh:384]: seems like a stray else?

Comment (ronen-fr) [reply qa/standalone/scrub/osd-scrub-repair.sh:384]: Indeed. I'm trying to understand how it got there after testing.

**Review by neha-ojha — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-19_06:38:14-rados-wip-kefu-testing-2021-05-19-1041-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-19_13:40:13-rados-wip-kefu-testing-2021-05-19-1831-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/50151
- https://tracker.ceph.com/issues/50891

**Pull request merged**